### PR TITLE
feat: introduce v3 schema with Go template support

### PR DESCRIPTION
**Pull request by fzipi**
## Summary

- Adds `RequestTemplate` type and a file-level `template` field to `FTWTest`, so tests can define the HTTP request structure once and inject per-test payloads via `{{.Payload}}` in URI, data, and header values
- Tests can now use a simplified format (`id` + `description` + `payload` + `output`) instead of explicit stages; the `stages` format remains fully supported for multi-step tests
- Bumps the Go module path to `github.com/coreruleset/ftw-tests-schema/v3`

## Changes

- New `RequestTemplate` struct with template-capable `uri`, `data`, and header values
- `FTWTest.Template` — file-level default request template
- `Test.Payload`, `Test.Template`, `Test.Output` — simplified per-test fields
- Custom `UnmarshalYAML` on `Test` accepts `id` / `description` as aliases for `test_id` / `desc`
- `spec/v3.0.0/` — generated JSON schemas for tests and platform overrides
- README updated with v3 and v2 format examples, version history table with comments

## Backward compatibility

All v2 test files parse and behave identically. Deprecated fields (`stop_magic`, `log_contains`, `no_log_contains`, `headers` map, `enabled`, `test_title`) are preserved in structs so existing test files produce correct results without modification.

## Test plan

- [ ] `go test ./...` passes (all 6 tests green)
- [ ] `mage jsonSchemas` regenerates `spec/v3.0.0/` correctly
- [ ] Existing v2-format YAML in `types/types_test.go` still unmarshals correctly (`TestUnmarshalFTWTest`)
- [ ] New v3-format YAML parses correctly, including `id`/`description` aliases (`TestUnmarshalV3Format`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Released FTW Tests YAML schema v3 with file-level default request templates, payload-based simplified test definitions, and field aliases (`id`, `description`) with backward-compatible parsing.
  * Added v3 JSON Schemas for core tests and platform overrides.
* **Documentation**
  * Extensively updated README to document schema v3 details (templates, overrides, stages behavior) and developer workflows.
* **Tests**
  * Added/extended tests covering v3 unmarshalling, alias handling, template/payload parsing, and expected stage behavior.

### CODE REVIEW

File: spec/v3.0.0/waf-tests-schema-v3.0.0.json [hunk 1]
```diff
@@ -0,0 +1 @@
...":{"type":"string"},"uri":{"type":"string"},"version":{"type":"string"},"method":{"type":"string"},"headers":{"items":{"$ref":"#/$defs/HeaderTuple"},"type":"array"},"data":{"type":"string"},"autocomplete_headers":{"type":"boolean"},"stop_magic":{"type":"boolean"},"save_cookie":{"type":"boolean"}},"additionalProperties":false,"type":"object"},"Response":{"properties":{"headers":{"additionalProperties":{"type":"string"},"type":"object"},"status":{"type":"integer"},"body":{"type":"string"},"encoded_body":{"type":"string"},"log_message":{"type":"string"}},"additionalProperties":false,"type":"object"},"Stage":{"properties":{"description":{"type":"string"},"input":{"$ref":"#/$defs/Input"},"output":{"$ref":"#/$defs/Output"}},"additionalProperties":false,"type":"object","required":["input","output"]},"Test":{"properties":{"test_title":{"type":"string"},"test_id":{"type":"integer"},"desc":{"type":"string"},"payload":{"type":"string"},"template":{"$ref":"#/$defs/RequestTemplate"},"output":{"$ref":"#/$defs/Output"},"stages":{"items":{"$ref":"#/$defs/Stage"},"type":"array"},"tags":{"items":{"type":"string"},"type":"array"}},"additionalProperties":false,"type":"object","required":["test_id"]}}}
```
[the remark this answers is not in the document]
Comment (fzipi) [reply spec/v3.0.0/waf-tests-schema-v3.0.0.json]: The `allOf` / `if-then` constraint for payload-mode vs stages-mode is intentionally deferred for the following reasons:

**Backward compatibility.** The v2 stage-based format places `output` inside each `stage`, not at the test level. A test can validly have neither a top-level `payload` nor a top-level `output`. Adding a constraint that requires `output` when `payload` is present — or disallows `stages` when `payload` is present — would break schema validation for mixed files that include both v2 and v3 tests.

**Validation is runtime, not schema-level.** The FTW engine already enforces this contract: if `stages` is populated it is used regardless of `payload`, and if `payload` is set with no `stages` the engine generates the stage from the file-level `template`. Duplicating that logic in JSON Schema adds coupling without adding safety that the engine does not already provide.

**allOf / if-then ordering edge cases.** The suggested constraint fires on the presence of the field in the serialised document. Because `stages` is `omitempty` it is absent from the serialised form of a payload-mode test, so the `if: {required: [stages]}` branch never triggers false positives — but the inverse (`if: {required: [payload]}`) would reject the large body of existing v2 test files that happen to include a `payload` key for reasons unrelated to the template mode.

This can be revisited in a future minor version once the v2 migration is complete and both formats no longer need to coexist in the same files.

**Comment by fzipi** [comment]
@coderabbitui full_review

File: types/types.go [hunk 2]
```diff
@@ -116,11 +147,50 @@ type Test struct {
 	//   - value: ExampleTest.TestDescription
 	TestDescription string `yaml:"desc,omitempty" json:"desc,omitempty"`
 
+	// description: |
+	//   Description is the v3 shorthand alias for desc. Either field may be used; if both are
```
Comment (theseion) [inline types/types.go]:
```suggestion
	//   Description is the v3 alias for desc. Either field may be used; if both are
```

File: types/types.go [hunk 3]
```diff
@@ -130,13 +200,146 @@ type Test struct {
 	Tags []string `yaml:"tags,omitempty" json:"tags,omitempty"`
 }
 
+// UnmarshalYAML implements yaml.Unmarshaler to accept v3 shorthand field names
+// `id` (alias for `test_id`) and `description` (alias for `desc`).
```
Comment (theseion) [inline types/types.go]:
```suggestion
// UnmarshalYAML parses a YAML document into a Test
```

File: types/types.go [hunk 4]
```diff
@@ -130,13 +200,146 @@ type Test struct {
...
+func (t *Test) UnmarshalYAML(value *yaml.Node) error {
```
Comment (theseion) [inline types/types.go:192]: Maybe we should add this to v2 as well. Not that we need it for go-ftw, but it's nice that tools don't need to implement it.

File: types/types.go [hunk 5]
```diff
@@ -130,13 +200,146 @@ type Test struct {
...
+	type rawTest struct {
```
Comment (theseion) [inline types/types.go]: Why do we need this struct? Can't we just pass in a struct of type `Test`? We already have everything defined.

File: types/types.go [hunk 6]
```diff
@@ -25,6 +29,18 @@ type FTWTest struct {
 	//     value: 123456
 	RuleId uint `yaml:"rule_id" json:"rule_id"`
 
+	// description: |
+	//   Template defines the default HTTP request structure for all payload-based tests in this file.
+	//
+	//   Field values in URI, Data, and header values may contain Go template expressions.
```
Comment (theseion) [inline types/types.go]:
```suggestion
	//   Field values in uri, data, and header values may contain Go template expressions.
```

File: README.md [hunk 7]
```diff
@@ -2,23 +2,220 @@
...
+
+The following template variables are available inside `uri`, `data`, and header values:
+
+| Variable | Description |
+|----------|-------------|
+| `{{.Payload}}` | The raw `payload` string from the test |
+
+### Per-test template override
+
+A test can override individual fields from the file-level template via its own `template` section. Only fields that are set in the per-test template take precedence; the rest fall through to the file-level template.
```
Comment (theseion) [inline README.md:41]: This should be documented in the previous section, it's doesn't really have anything to do with template variables.

File: README.md [hunk 8]
```diff
@@ -2,23 +2,220 @@
...
+      data: "q={{.Payload}}"
+    payload: "1 UNION SELECT 1,2,3--"
+    output:
+      log:
+        expect_ids: [942100]
+```
+
+### Field name aliases
+
+For a cleaner authoring experience, v3 accepts short aliases for two common fields:
```
Comment (theseion) [inline README.md]:
```suggestion
For a cleaner authoring experience, v3 accepts aliases for two common fields:
```

File: README.md [hunk 9]
```diff
@@ -2,23 +2,220 @@
...
+
+| v3 alias | Canonical field |
+|----------|----------------|
+| `id` | `test_id` |
+| `description` | `desc` |
```
Comment (theseion) [inline README.md:72]: TBH, I would deprecate `test_id` and `desc`. Then the v3 fields become the canonical form and the deprecated names are the aliases.

**Review by theseion — CHANGES_REQUESTED** [review]
❤️

File: types/types.go (hunk 4 above)
Comment (fzipi) [reply types/types.go:192]: Good point. Keeping it v3-only for now since v2 files already parse correctly with `test_id`/`desc`. Happy to add the aliases to v2 in a follow-up if there's demand from tooling authors.

File: types/types.go (hunk 5 above)
Comment (fzipi) [reply types/types.go]: Done in 604b204 — refactored to `type rawTest Test` (strips methods to avoid infinite recursion) with a small `legacy` struct just for `test_id` and `desc`.

File: README.md (hunk 7 above)
Comment (fzipi) [reply README.md:41]: Done in 604b204 — moved the per-test override section before Template variables.

File: README.md (hunk 9 above)
Comment (fzipi) [reply README.md:72]: Done in 604b204 — `id` and `description` are now the canonical v3 YAML/JSON names; `test_id` and `desc` are deprecated aliases accepted by `UnmarshalYAML` for backward compatibility.

**Review by theseion — APPROVED** [review]
